src/interpreter/frame.rs

```diff
@@ -0,0 +1,923 @@
+use rustc::hir::def_id::DefId;
+use rustc::mir::repr as mir;
+use rustc::ty::layout::{self, Layout, Size};
+use rustc::ty::subst::{self, Subst, Substs};
+use rustc::ty::{self, Ty};
+use std::{iter, mem};
+use syntax::attr;
+use syntax::codemap::{self, Span};
+
+use error::{EvalError, EvalResult};
+use memory::Pointer;
+use primval::{self, PrimVal};
+
+use interpreter::global::GlobalEvalContext;
+
+use super::{
+    CachedMir,
+    ConstantId,
+    Lvalue,
+    LvalueExtra,
+    pointee_type,
+    StructExt,
+    IntegerExt,
+    TerminatorTarget,
+};
+
+/// A stack frame.
+pub(super) struct Frame<'mir, 'tcx: 'mir> {
+    /// The def_id of the current function
+    pub def_id: DefId,
+
+    /// The span of the call site
+    pub span: codemap::Span,
+
+    /// type substitutions for the current function invocation
+    pub substs: &'tcx Substs<'tcx>,
+
+    /// The MIR for the function called on this frame.
+    pub mir: CachedMir<'mir, 'tcx>,
+
+    /// The block that is currently executed (or will be executed after the above call stacks return)
+    pub next_block: mir::BasicBlock,
+
+    /// A pointer for writing the return value of the current call if it's not a diverging call.
+    pub return_ptr: Option<Pointer>,
+
+    /// The list of locals for the current function, stored in order as
+    /// `[arguments..., variables..., temporaries...]`. The variables begin at `self.var_offset`
+    /// and the temporaries at `self.temp_offset`.
+    pub locals: Vec<Pointer>,
+
+    /// The offset of the first variable in `self.locals`.
+    pub var_offset: usize,
+
+    /// The offset of the first temporary in `self.locals`.
+    pub temp_offset: usize,
+
+    /// The index of the currently evaluated statment
+    pub stmt: usize,
+
+    // Constants that need to be evaluated before the next statement can be evaluated
+    pub constants: Vec<(ConstantId<'tcx>, Span, Pointer, CachedMir<'mir, 'tcx>)>,
+}
+
+pub(super) struct FrameInterpreter<'a, 'b: 'a, 'mir: 'a, 'tcx: 'a + 'mir + 'b> {
```

Comment:
I don't understand the `FrameInterpreter`/`FnEvalContext` split. Why is statement evaluation here and terminator evaluation over there?

I want to lean towards fewer "context" types if possible. I wasn't really happy I had to make the `Global`/`Fn` split, either.